### PR TITLE
Initial payment store setup

### PR DESCRIPTION
**Pull request by danielgranhao**
Handles persisting new receiving payments (when invoice is created), filling the preimage when we learn about it, and marking a receiving payment as successful when the funds are claimed.

### CODE REVIEW

File: eel/src/payment_store.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        .execute_batch(
+            "\
+            CREATE TABLE IF NOT EXISTS payments (
+              payment_id integer NOT NULL PRIMARY KEY,
+              type text CHECK( type IN ('receiving', 'sending') ) NOT NULL,
+              hash tinyblob NOT NULL,
+              amount_msat bigint NOT NULL,
+              invoice text NOT NULL,
+              preimage tinyblob,
+              fees_msat bigint,
```
[the remark this answers is not in the document]
Comment (danielgranhao) [reply eel/src/payment_store.rs]: Addressed in [0e56ea5](https://github.com/getlipa/lipa-lightning-lib/pull/212/commits/0e56ea5c1873b46e1d7ef4a2b12b7017ea6d62ad)

File: eel/src/payment_store.rs [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+              type text CHECK( type in ('created', 'succeeded', 'failed') ) NOT NULL,
+              time timestamp NOT NULL,
+              current_fiat_value real NOT NULL,
+              FOREIGN KEY (payment_id) REFERENCES payments(payment_id)
+            );
+        ",
+        )
+        .map_to_permanent_failure("Failed to set up local payment database")?;
+
+    Ok(())
```
[the remark this answers is not in the document]
Comment (danielgranhao) [reply eel/src/payment_store.rs]: Fixed in [8ab8cfe](https://github.com/getlipa/lipa-lightning-lib/pull/212/commits/8ab8cfe1ded24981fab72babbef433d564311ad6)

File: eel/src/payment_store.rs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+              preimage tinyblob,
+              network_fees_msat bigint,
+              lsp_fees_msat bigint,
+              metadata blob
+            );
+            CREATE TABLE IF NOT EXISTS events (
+              event_id integer NOT NULL PRIMARY KEY,
+              payment_id integer NOT NULL,
+              type text CHECK( type in ('created', 'succeeded', 'failed') ) NOT NULL,
+              time timestamp NOT NULL,
```
[the remark this answers is not in the document]
Comment (danielgranhao) [reply eel/src/payment_store.rs]: As far as I understand, if the timestamp is UTC, we shouldn't have problems regarding timezones. I've now changed to using SQL's `CURRENT_TIMESTAMP`, which I think also is UTC. We have to confirm, but if it isn't, we can simply go back to inserting `chrono::offset::Utc::now()`.

File: eel/src/payment_store.rs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    let metadata = row.get(8)?;
+    let payment_state: String = row.get(9)?;
+    let payment_state = match payment_state.as_str() {
+        "created" => PaymentState::Created,
+        "succeeded" => PaymentState::Succeeded,
+        "failed" => PaymentState::Failed,
+        _ => return Err(rusqlite::Error::ExecuteReturnedResults),
+    };
+    let timestamp: chrono::DateTime<chrono::Utc> = row.get(10)?;
+    let timestamp = SystemTime::from(timestamp);
```
[the remark this answers is not in the document]
Comment (danielgranhao) [reply eel/src/payment_store.rs:160]: The `rusqlite` crate implements the traits `FromSql` and `ToSql` for certain types, making it easy to convert in either direction. The date/time types it implements are from `chrono`.

File: eel/src/payment_store.rs [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        network_fees_msat,
+        lsp_fees_msat,
+        metadata,
+    })
+}
+
+fn apply_migrations(db_conn: &Connection) -> Result<()> {
+    db_conn
+        .execute_batch(
+            "\
```
[the remark this answers is not in the document]
Comment (danielgranhao) [reply eel/src/payment_store.rs:178]: Applied in [1ccbe3a](https://github.com/getlipa/lipa-lightning-lib/pull/212/commits/1ccbe3a51b0b210c6425c4119963847a8e331c23)

File: eel/src/payment_store.rs [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+              network_fees_msat bigint,
+              lsp_fees_msat bigint,
+              metadata blob
+            );
+            CREATE TABLE IF NOT EXISTS events (
+              event_id integer NOT NULL PRIMARY KEY,
+              payment_id integer NOT NULL,
+              type text CHECK( type in ('created', 'succeeded', 'failed') ) NOT NULL,
+              time timestamp NOT NULL,
+              current_fiat_value real NOT NULL,
```
[the remark this answers is not in the document]
Comment (danielgranhao) [reply eel/src/payment_store.rs]: Humm.. yes, we need to find a way to solve that. Maybe store all supported fiat conversion rates as metadata? For now, I completely removed the fiat part in [1ccbe3a](https://github.com/getlipa/lipa-lightning-lib/pull/212/commits/1ccbe3a51b0b210c6425c4119963847a8e331c23)

File: eel/src/payment_store.rs [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+              preimage tinyblob,
+              network_fees_msat bigint,
+              lsp_fees_msat bigint,
+              metadata blob
+            );
+            CREATE TABLE IF NOT EXISTS events (
+              event_id integer NOT NULL PRIMARY KEY,
+              payment_id integer NOT NULL,
+              type text CHECK( type in ('created', 'succeeded', 'failed') ) NOT NULL,
+              time timestamp NOT NULL,
```
[the remark this answers is not in the document]
Comment (danielgranhao) [reply eel/src/payment_store.rs]: Applied in [1ccbe3a](https://github.com/getlipa/lipa-lightning-lib/pull/212/commits/1ccbe3a51b0b210c6425c4119963847a8e331c23)

File: eel/examples/node/cli.rs [hunk 8]
```diff
@@ -182,3 +191,32 @@ fn pay_invoice<'a>(
...
     Ok(())
 }
+
+fn list_payments(node: &LightningNode) -> Result<(), String> {
+    let payments = match node.get_latest_payments(100) {
+        Ok(p) => p,
+        Err(e) => return Err(e.to_string()),
+    };
+
+    println!("Total of {} payments\n", payments.len());
```
Comment (gcomte) [inline eel/examples/node/cli.rs:201]: What information are you wanting to log here, if the payments length is hardcoded to be `100` ?

Comment (danielgranhao) [reply eel/examples/node/cli.rs:201]: The idea was to provide the user of the CLI with the number of payments that will be printed immediately after. But we can improve the example node for this command to accept the number of payments, making the limit more understandable.

Comment (gcomte) [reply eel/examples/node/cli.rs:201]: Yes maybe later, I think it is good enough for now.

**Pull request merged**